routesrv/routesrv_test.go

```diff
@@ -0,0 +1,223 @@
+	logrus.AddHook(tl)
+	os.Exit(m.Run())
+}
+
+func TestNotInitializedRoutesAreNotServed(t *testing.T) {
+	defer tl.Reset()
+	ks, _ := newKubeServer(t)
+	rs := newRouteServer(t, ks)
+
+	w := getRoutes(rs)
+
+	if len(w.Body.Bytes()) > 0 {
+		t.Error("uninitialized routes were served")
+	}
+	wantHTTPCode(t, w, http.StatusNotFound)
+}
+
+func TestEmptyRoutesAreNotServed(t *testing.T) {
+	defer tl.Reset()
+	ks, _ := newKubeServer(t)
+	ks.Start()
+	defer ks.Close()
+	rs := newRouteServer(t, ks)
+
+	rs.StartUpdates()
+	if err := tl.WaitFor(routesrv.LogRoutesEmpty, waitTimeout); err != nil {
+		t.Error("empty routes not received")
+	}
+	w := getRoutes(rs)
+
+	if len(w.Body.Bytes()) > 0 {
+		t.Error("empty routes were served")
+	}
+	wantHTTPCode(t, w, http.StatusNotFound)
+}
+
+func TestFetchedRoutesAreServedInEskipFormat(t *testing.T) {
+	defer tl.Reset()
+	ks, _ := newKubeServer(t, loadKubeYAML(t, "testdata/lb-target-multi.yaml"))
+	ks.Start()
+	defer ks.Close()
+	rs := newRouteServer(t, ks)
+
+	rs.StartUpdates()
+	if err := tl.WaitFor(routesrv.LogRoutesInitialized, waitTimeout); err != nil {
+		t.Error("routes not initialized")
+	}
+	w := getRoutes(rs)
+
+	want := parseEskipFixture(t, "testdata/lb-target-multi.eskip")
+	got, err := eskip.Parse(w.Body.String())
+	if err != nil {
+		t.Errorf("served routes are not valid eskip: %s", w.Body)
+	}
+	if !eskip.EqLists(got, want) {
+		t.Errorf("served routes do not reflect kubernetes resources: %s", cmp.Diff(got, want))
+	}
+	wantHTTPCode(t, w, http.StatusOK)
+}
+
+func TestLastRoutesAreServedDespiteSourceFailure(t *testing.T) {
+	defer tl.Reset()
+	ks, _ := newKubeServer(t, loadKubeYAML(t, "testdata/lb-target-multi.yaml"))
+	ks.Start()
+	defer ks.Close()
+	rs := newRouteServer(t, ks)
+
+	rs.StartUpdates()
+	if err := tl.WaitFor(routesrv.LogRoutesInitialized, waitTimeout); err != nil {
+		t.Error("routes not initialized")
+	}
+	w1 := getRoutes(rs)
+	wantHTTPCode(t, w1, http.StatusOK)
+
+	ks.Close()
+	if err := tl.WaitFor(routesrv.LogRoutesFetchingFailed, waitTimeout); err != nil {
+		t.Error("source failure not recognized")
+	}
+	w2 := getRoutes(rs)
+
+	if !bytes.Equal(w1.Body.Bytes(), w2.Body.Bytes()) {
+		t.Error("served routes changed after source failure")
+	}
+	wantHTTPCode(t, w2, http.StatusOK)
+}
+
+func TestRoutesAreUpdated(t *testing.T) {
+	defer tl.Reset()
+	ks, handler := newKubeServer(t, loadKubeYAML(t, "testdata/lb-target-multi.yaml"))
+	ks.Start()
+	defer ks.Close()
+	rs := newRouteServer(t, ks)
+
+	rs.StartUpdates()
+	if err := tl.WaitFor(routesrv.LogRoutesInitialized, waitTimeout); err != nil {
+		t.Error("routes not initialized")
+	}
+	w1 := getRoutes(rs)
+
+	handler.set(newKubeAPI(t, loadKubeYAML(t, "testdata/lb-target-single.yaml")))
+	if err := tl.WaitForN(routesrv.LogRoutesUpdated, 2, waitTimeout*2); err != nil {
+		t.Error("routes not updated")
+	}
+	w2 := getRoutes(rs)
+
+	if bytes.Equal(w1.Body.Bytes(), w2.Body.Bytes()) {
+		t.Error("route contents were not updated")
+	}
+}
```

Comment:
great tests!
What about adding one that tests that we actually update in case of w1 is different from w2?

Reply:
Maybe I misunderstood your message but I think this is exactly this test:
1. We start the API server populated with resources (`testdata/lb-target-multi.yaml`) and polling.
2. We make sure the routes corresponding to the API resources were initialized.
3. We **change** what we serve with the API server (`testdata/lb-target-single.yaml`).
4. We make sure the new routes corresponding to the new API resources were updated.
5. We assert if we actually served something new (expecting the change, because of the change in the API server) by checking if `w1` is different from `w2`.

Maybe you can rephrase or clarify? I am also open for a quick call to do that.